### PR TITLE
Handle PageUp/PageDown for completion 

### PR DESCRIPTION
**Pull request by Bensmuth**
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/4301)

### CODE REVIEW

**Comment by The-Compiler** [comment]
Thanks! The idea there was that PageUp/PageDown go up/down for multiple items (similar how it'd behave in a list of items in another application, or in a text editor). That means you'd need to modify the `:completion-focus` command to handle something like `next-page` and `prev-page`. Do you want to give that a try?

**Comment by Kenoben** [comment]
Yeah sure I'll get right to it

**Comment by The-Compiler** [comment]
Hardcoding 14 items unfortunately won't work out, as the completion height differs (depending on the config or the window size).

Instead, you might want to use [visualRect](http://doc.qt.io/qt-5/qabstractitemview.html#visualRect) to find out the size of a single item (and calculate how many items are displayed based on that), or perhaps [indexAt](http://doc.qt.io/qt-5/qabstractitemview.html#indexAt) to get the index of the item at the bottom of the screen.

This might be a bit more involved than I originally thought, as it also involves multiple coordinate systems and stuff - unfortunately, I'll be unable to help much currently (see [this announcement](https://lists.schokokeks.org/pipermail/qutebrowser-announce/2018-September/000051.html) - I should really get started on that, currently trying to review remaining open PRs). Maybe @rcorre has some ideas how to proceed, but if you think that's getting too involved, feel free to pick something else. Sorry for the trouble!

**Comment by Kenoben** [comment]
I'll give it a shot

**Comment by Kenoben** [comment]
So hopefully this fixes the issue, I used some simple maths to do it all (Line 209), the only issue I can see is the use of floats instead of integers in the algorithm.

**Comment by jgkamat** [comment]
You might be able to use [rowHeight](http://doc.qt.io/qt-5/qtreeview.html#rowHeight), but I don't know if that's the best way to do it.

**Comment by rcorre** [comment]
> Instead, you might want to use visualRect to find out the size of a single item (and calculate how many items are displayed based on that), or perhaps indexAt to get the index of the item at the bottom of the screen.

Yup, I would think something like this. Don't really have anything clever to add ;)

**Comment by Kenoben** [comment]
This should implement the requested feature, I've tested it with different font sizes in :set fonts.completion.entry and it worked.
Sorry if I've been a massive pain.

**Comment by The-Compiler** [comment]
Looks like tests are currently [failing](https://travis-ci.org/qutebrowser/qutebrowser/jobs/438241875#L604) because of a `ZeroDivisonError`. I'm not sure why the height there would be zero - @rcorre do you have an idea?

File: qutebrowser/completion/completionwidget.py [hunk 1]
```diff
@@ -196,6 +196,47 @@ def _next_idx(self, upwards):
...
+                return self.model().last_item()
+            else:
+                return self.model().first_item()
+        else:
+            element_height = self.visualRect(idx).height()  # Finds Height of each CompletionView elements
+            page_length = int(self.height()/element_height)
+
+        while True:
+            if upwards:
+                for x in range(page_length-1):  # Ensures bottom element stays at top of next page
```
Comment (rcorre) [inline qutebrowser/completion/completionwidget.py]: use `_` for an unused loop variable

File: qutebrowser/completion/completionwidget.py [hunk 2]
```diff
@@ -196,6 +196,47 @@ def _next_idx(self, upwards):
...
+
+        if not idx.isValid():
+            # No item selected yet
+            page_length = 0  # No need to scroll yet
+            if upwards:
+                return self.model().last_item()
+            else:
+                return self.model().first_item()
+        else:
+            element_height = self.visualRect(idx).height()  # Finds Height of each CompletionView elements
```
Comment (rcorre) [inline qutebrowser/completion/completionwidget.py]: put these comments on the line above to limit line length

File: qutebrowser/completion/completionwidget.py [hunk 3]
```diff
@@ -196,6 +196,47 @@ def _next_idx(self, upwards):
 
         raise utils.Unreachable
 
+    def _next_page(self, upwards):
+        """Essentially _next_idx with a for loop to increment by amount of elements per page (size of each page) instead of 1 each time
```
Comment (rcorre) [inline qutebrowser/completion/completionwidget.py]: wrap the line here, though I think a simpler summary would do.
Maybe "Return the index a page away from the selected index".
Mentioning implementation details like `with a for loop` shouldn't be necessary.

File: qutebrowser/completion/completionwidget.py [hunk 4]
```diff
@@ -263,6 +304,8 @@ def completion_item_focus(self, which, history=False):
             'prev': self._next_idx(upwards=True),
```
Comment (rcorre) [inline qutebrowser/completion/completionwidget.py]: You were just following the pattern here, but I think these should all be changed to lambdas. Right now it is calculating all of the indices and only using one of them. That is why the [test is failing](https://travis-ci.org/qutebrowser/qutebrowser/jobs/438241875#L702) when you haven't even added tests for 'next/last-page' yet.

IMHO an `if/elif` block would be fine too.

That being said, we still need to figure out why `self.visualRect(idx).height()` is sometimes `0`.

Comment (rcorre) [reply qutebrowser/completion/completionwidget.py]: Maybe you just need to set a font or something for the tests? I can't get it to fail when actually running the browser.

Comment (Kenoben) [reply qutebrowser/completion/completionwidget.py]: Sorry, but what lines do you want implemented as lambdas?

Comment (rcorre) [reply qutebrowser/completion/completionwidget.py]:
```py
    indices = {
        'next': lambda: self._next_idx(upwards=False),
        'prev': lambda: self._next_idx(upwards=True),
        'next-category': lambda: self._next_category_idx(upwards=False),
        'prev-category': lambda: self._next_category_idx(upwards=True),
        'next-page': lambda: self._next_page(upwards=False),
        'last-page': lambda: self._next_page(upwards=True),
    }
    idx = indices[which]()
```

or

```py
if which == 'next':
    idx = self._next_idx(upwards=False)
elif which == 'prev':
    idx = self._next_idx(upwards=True)
elif which == 'next-category':
    idx = self._next_category_idx(upwards=False)
elif which == 'prev-category':
    idx = self._next_category_idx(upwards=True)
elif which == 'next-page':
    idx = self._next_page(upwards=False)
elif which == 'last-page':
    idx = self._next_page(upwards=True)
```

So we aren't computing every index when we only want one.

**Comment by rcorre** [comment]
http://doc.qt.io/qt-5/qtreeview.html#visualRect: 

> If the index is not visible or explicitly hidden, the returned rectangle is invalid.

```
            element_height = self.visualRect(idx).height()  # Finds Height of each CompletionView elements
            print(idx.isValid(), self.visualRect(idx).isValid(), element_height)
            page_length = int(self.height()/element_height)
```

I'm getting `True False 0` for the above `print` during the tests, so I'm guessing index isn't visible.

File: qutebrowser/completion/completionwidget.py [hunk 5]
```diff
@@ -216,15 +216,17 @@ def _next_page(self, upwards):
...
 
         while True:
             if upwards:
-                for x in range(page_length-1):  # Ensures bottom element stays at top of next page
+                # Ensures bottom element stays at top of next page
+                for _x in range(page_length-1):  
                     idx = self.indexAbove(idx)
             else:
-                for x in range(page_length-1):
+                for _x in range(page_length-1):
```
Comment (rcorre) [inline qutebrowser/completion/completionwidget.py]: just `for _ in range`. That seems to be the most established convention in this repo.

**Comment by rcorre** [comment]
Looks good! Take a look at the [pylint errors](https://travis-ci.org/qutebrowser/qutebrowser/jobs/441027047#L489), add some test cases for pageup/pagedown, and you should be good to go!

**Comment by arza-zara** [comment]
I fixed wrapping, scroll position when jumping to category header, and code style. How should tests be written?

**Comment by The-Compiler** [comment]
This is one of those occurrences where something seems quite easy, but then it turns out there are a lot more corner cases around than you'd have thought. Merged with some tests and a couple of simplifications:

- 6b3130ee7 Merge remote-tracking branch 'origin/pr/4301' into pageupdown
- 6bba912fe tests: Use fixture for model in CompletionWidget tests
- 2dff1c43c Add tests for :completion-item-focus with prev-page/next-page
- debf8f702 Simplify CompletionView._next_page a bit
- 156f3a66c Calculate index instead of looping
- e8b78de67 Update docs/changelog

Thanks @Bensmuth for the PR, @arza-zara for the follow-up, and @rcorre and @jgkamat for the comments!

Also, a late answer to:

> Also, I somewhat disagree with the style rule here, and that isn't required in the style guides either: [...]

That is mentioned (though not required) in [PEP 8](https://www.python.org/dev/peps/pep-0008/#indentation):

> When the conditional part of an if-statement is long enough to require that it be written across multiple lines, it's worth noting that the combination of a two character keyword (i.e. if), plus a single space, plus an opening parenthesis creates a natural 4-space indent for the subsequent lines of the multiline conditional. This can produce a visual conflict with the indented suite of code nested inside the if-statement, which would also naturally be indented to 4 spaces. This PEP takes no explicit position on how (or whether) to further visually distinguish such conditional lines from the nested suite inside the if-statement. Acceptable options in this situation include, but are not limited to:
>
> ```python
> # No extra indentation.
> if (this_is_one_thing and
>     that_is_another_thing):
>     do_something()
> 
> # Add a comment, which will provide some distinction in editors
> # supporting syntax highlighting.
> if (this_is_one_thing and
>     that_is_another_thing):
>     # Since both conditions are true, we can frobnicate.
>     do_something()
> 
> # Add some extra indentation on the conditional continuation line.
> if (this_is_one_thing
>         and that_is_another_thing):
>     do_something()
> ```

So far, qutebrowser consistently uses the third form, and I think I'd prefer to keep that.
